File: app/src/main/java/com/myapp/lexicon/service/LexiconService.java
package com.myapp.lexicon.service;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.IBinder;
import android.preference.PreferenceManager;
import android.widget.Toast;

import com.myapp.lexicon.R;
import com.myapp.lexicon.database.AppDB;
import com.myapp.lexicon.database.DataBaseEntry;
import com.myapp.lexicon.database.DatabaseHelper;
import com.myapp.lexicon.schedule.AppNotification;
import com.myapp.lexicon.settings.AppData;
import com.myapp.lexicon.settings.AppSettings;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import kotlin.Pair;

public class LexiconService extends Service implements ServiceDialog.IStopServiceByUser
{
    public static boolean stopedByUser = false;
    private Locale oldLocale;
    private PhoneUnlockedReceiver receiver;
    private int displayVariant = 0;
    private int startId;
    private final int appId = 542390;
    NotificationManager manager;

    public LexiconService()
    {
    }

    @Override
    public IBinder onBind(Intent intent)
    {
        return null;
    }

    @Override
    public void onCreate()
    {
        super.onCreate();
        oldLocale = getResources().getConfiguration().locale;

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(LexiconService.this);
        String preferencesString = preferences.getString(getString(R.string.key_display_variant), "0");
        displayVariant = Integer.parseInt(preferencesString);

        String contentText = "";
        Intent intent;
        PendingIntent pendingIntent = null;
        Notification appNotification;
        switch (displayVariant)
        {
            case 0:
                contentText = getString(R.string.notify_content_text);
                break;
            case 1:
                contentText = getString(R.string.notify_new_word_text);
                break;
        }

        appNotification = new AppNotification(this).create(new DataBaseEntry(contentText, ""));
        startForeground(AppNotification.NOTIFICATION_ID, appNotification);

        receiver = new PhoneUnlockedReceiver();
        IntentFilter filter = new IntentFilter();
        filter.addAction(Intent.ACTION_USER_PRESENT);
        filter.addAction(Intent.ACTION_SCREEN_OFF);
        registerReceiver(receiver, filter);

        ServiceDialog.setStoppedByUserListener(this);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId)
    {
        this.startId = startId;
        return START_STICKY;
    }

    @Override
    public boolean onUnbind(Intent intent)
    {
        return super.onUnbind(intent);
    }

    @Override
    public void onDestroy()
    {
        unregisterReceiver(receiver);
        super.onDestroy();
    }

    @Override
    public void onTaskRemoved(Intent rootIntent)
    {
        super.onTaskRemoved(rootIntent);
        onCreate();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig)
    {
        super.onConfigurationChanged(newConfig);
        Locale newLocale = newConfig.locale;
        if (!oldLocale.getLanguage().equals(newLocale.getLanguage()))
        {
            AppSettings appSettings = new AppSettings(this);
            appSettings.cleanPlayList();
            stopSelf(startId);
        }
    }

    @Override
    public void onStoppedByUser()
    {
        stopSelf(startId);
        Toast.makeText(this, this.getString(R.string.text_app_is_closed) + " " + getString(R.string.app_name) + " " + this.getString(R.string.text_app_is_closed_end), Toast.LENGTH_SHORT).show();
    }




    public class PhoneUnlockedReceiver extends BroadcastReceiver
    {
        // TODO: обработчик событий нажатия кнопки блокировки, выключения экрана....
        @Override
        public void onReceive(Context context, Intent intent)
        {
            if (context != null && intent != null)
            {
                SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
                String displayVariant = preferences.getString(context.getString(R.string.key_display_variant), "0");
                String displayMode = preferences.getString(context.getString(R.string.key_list_display_mode), "0");

                String action = intent.getAction();
                //String actionUserPresent = Intent.ACTION_USER_PRESENT;
                String actionScreenOff = Intent.ACTION_SCREEN_OFF;

                if (action != null)
                {
                    if ((action.equals(actionScreenOff) /*|| action.equals(actionUserPresent)*/))
                    {
                        if (displayVariant.equals("0"))
                        {
                            Intent intentAct = new Intent(context, ServiceDialog.class);
                            intentAct.setAction(Intent.ACTION_MAIN);
                            intentAct.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                            startActivity(intentAct);
                        }
                        if (displayVariant.equals("1"))
                        {
                            AppData appData = AppData.getInstance();
                            ArrayList<String> playList = appData.getPlayList();
                            String dictName = playList.get(appData.getNdict());
                            AppDB db = new AppDB(new DatabaseHelper(context));

                            db.getEntriesAndAmountAsync(dictName, appData.getNword(), "ASC")
                                    .subscribeOn(Schedulers.computation())
                                    .observeOn(AndroidSchedulers.mainThread())
                                    .subscribe(item -> {
                                        Pair<Map<String, Integer>, List<DataBaseEntry>> pairs = item;
                                        if (displayMode.equals("0"))
                                        {
                                            if (pairs.getSecond().size() > 0)
                                            {
                                                AppNotification appNotification = new AppNotification(context);
                                                appNotification.create(pairs.getSecond().get(0));
                                                appNotification.show();
                                            }
                                            if (pairs.getSecond().size() > 1)
                                            {
                                                appData.setNword(pairs.getSecond().get(1).getRowId());
                                            }
                                            if (pairs.getSecond().size() == 1)
                                            {
                                                appData.setNword(1);
                                                if (appData.getNdict() >=0 && appData.getNdict() <= playList.size() - 2)
                                                {
                                                    appData.setNdict(appData.getNdict() + 1);
                                                }
                                                else appData.setNdict(0);
                                            }
                                        }
                                        if (displayMode.equals("1"))
                                        {
                                            if (pairs.getSecond().size() > 0)
                                            {
                                                AppNotification appNotification = new AppNotification(context);
                                                appNotification.create(pairs.getSecond().get(0));
                                                appNotification.show();
                                            }
                                        }
                                    }, Throwable::printStackTrace);

                        }
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/myapp/lexicon/service/LexiconService.kt
package com.myapp.lexicon.service

import android.annotation.SuppressLint
import android.app.Service
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.IBinder
import android.widget.Toast
import androidx.core.content.ContextCompat.startActivity
import androidx.preference.PreferenceManager
import com.google.gson.Gson
import com.myapp.lexicon.R
import com.myapp.lexicon.database.AppDataBase
import com.myapp.lexicon.models.Word
import com.myapp.lexicon.repository.DataRepositoryImpl
import com.myapp.lexicon.schedule.AppNotification
import com.myapp.lexicon.service.ServiceActivity.Listener
import com.myapp.lexicon.settings.AppSettings
import com.myapp.lexicon.settings.getWordFromPref
import com.myapp.lexicon.settings.saveWordToPref
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.disposables.CompositeDisposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


@AndroidEntryPoint
class LexiconService : Service(), Listener
{
    private var blockReceiver: PhoneUnlockedReceiver? = null
    private var startId = 0
    private val composite = CompositeDisposable()
    private lateinit var repository: DataRepositoryImpl

    override fun onBind(intent: Intent): IBinder?
    {
        return null
    }

    override fun onCreate()
    {
        super.onCreate()

        val dao = AppDataBase.buildDataBase(this).appDao()
        val appSettings = AppSettings(this)
        repository = DataRepositoryImpl(dao, appSettings)

        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val isService = preferences.getBoolean(getString(R.string.key_service), false)
        if (isService)
        {
            blockReceiver = PhoneUnlockedReceiver()
            val filter = IntentFilter()
            filter.addAction(Intent.ACTION_USER_PRESENT)
            filter.addAction(Intent.ACTION_SCREEN_OFF)
            registerReceiver(blockReceiver, filter)
        }

        ServiceActivity.setStoppedByUserListener(this)

        this.getWordFromPref(
            onInit = {
                CoroutineScope(Dispatchers.IO).launch {
                    val word = repository.getFirstEntryAsync().await()
                    showAppNotification(word)
                }
            },
            onSuccess = { word ->
                showAppNotification(word)
            },
            onFailure = {
                stopSelf()
            }
        )

    }

    private fun showAppNotification(word: Word) {
        CoroutineScope(Dispatchers.IO).launch {
            val words = repository.getEntriesByDictNameAsync(word.dictName, word._id.toLong(), 1, 2).await()
            if (words.isNotEmpty()) {
                val json = Gson().toJson(words)
                json?.let {
                    val appNotification = AppNotification(this@LexiconService).create(json)
                    //startForeground(AppNotification.NOTIFICATION_ID, appNotification)

                }
            }
        }
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int
    {
        this.startId = startId
        return START_STICKY
    }

    override fun onDestroy()
    {
        super.onDestroy()
        blockReceiver?.let { unregisterReceiver(blockReceiver) }
        composite.apply {
            dispose()
            clear()
        }
    }

    override fun onTaskRemoved(rootIntent: Intent)
    {
        super.onTaskRemoved(rootIntent)
        onCreate()
    }

    override fun onStoppedByUser()
    {
        stopSelf(startId)
        Toast.makeText(this, this.getString(R.string.text_app_is_closed) + " " + getString(R.string.app_name) + " " + this.getString(R.string.text_app_is_closed_end), Toast.LENGTH_SHORT).show()
    }



    // TODO: обработчик событий нажатия кнопки блокировки, выключения экрана....
    @AndroidEntryPoint
    class PhoneUnlockedReceiver : BroadcastReceiver()
    {
        private lateinit var repository: DataRepositoryImpl

        @SuppressLint("CheckResult")
        @Suppress("RedundantSamConstructor")
        override fun onReceive(context: Context, intent: Intent)
        {
            val dao = AppDataBase.buildDataBase(context).appDao()
            val appSettings = AppSettings(context)
            repository = DataRepositoryImpl(dao, appSettings)

            val preferences = PreferenceManager.getDefaultSharedPreferences(context)
            val displayVariant = preferences.getString(context.getString(R.string.key_display_variant), "0")!!
            val displayMode = preferences.getString(context.getString(R.string.key_list_display_mode), "0")!!
            val action = intent.action

            //String actionUserPresent = Intent.ACTION_USER_PRESENT;
            val actionScreenOff = Intent.ACTION_SCREEN_OFF
            if (action != null && action == actionScreenOff)
            {
                context.getWordFromPref(
                    onInit = {
                        CoroutineScope(Dispatchers.IO).launch {
                            val word = repository.getFirstEntryAsync().await()
                            handleBroadCast(context, word, displayVariant, displayMode)
                        }
                    },
                    onSuccess = { word ->
                        handleBroadCast(context, word, displayVariant, displayMode)
                    },
                    onFailure = {
                        it.printStackTrace()
                    }
                )
            }
        }

        private fun handleBroadCast(context: Context, word: Word, displayVariant: String, displayMode: String) {

            CoroutineScope(Dispatchers.IO).launch {
                val words = repository.getEntriesByDictNameAsync(word.dictName, id = word._id.toLong(), limit = 2).await()
                if (displayVariant == "0")
                {
                    val intentAct = Intent(context, ServiceActivity::class.java).apply {
                        action = Intent.ACTION_MAIN
                        addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
                        val json = Gson().toJson(words)
                        putExtra(ServiceActivity.ARG_JSON, json)
                    }
                    startActivity(context, intentAct, Bundle.EMPTY)
                }
                if (displayVariant == "1")
                {
                    val json = Gson().toJson(words)
                    val appNotification = AppNotification(context)
                    appNotification.create(json)
                    appNotification.show()
                    if (displayMode == "0")
                    {
                        when(words.size) {
                            2 -> {
                                context.saveWordToPref(words[1])
                            }
                            1 -> {
                                context.saveWordToPref(words[0])
                            }
                        }
                    }
                }
            }

        }

    }



}